### PR TITLE
Add `SubCircuit::unusable_rows` and tests for each `SubCircuit`

### PR DESCRIPTION
**Pull request by han0110**
### Description

To remove magic numbers for blinding factors, this PR adds an associated function `unusable_rows` for `SubCircuit` to returns their `unusable_rows`, which should be `meta.blinding_factors() + 1`.

Tests are also added to make sure the returned values are correct. 

Note that currently `KeccakCircuit` could be configured by environment variable `KECCAK_ROWS` to decide how many rows a round would take, which affect the different rotation queried, and it's unfortunately not a easy way to compute how many different rotations are used, so this PR hardcodes the unusable rows from `KECCAK_ROWS = 1` to `20` for easy lookup.

### Issue Link

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/949

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

[_explanation_]

### CODE REVIEW

File: zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs
Comment (CPerezz) [inline zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:25]: Where does this rows info come from? Why are all keccak-rows different in terms of unusable rows?

**Review by CPerezz — COMMENTED** [review]
Looks good overall. 

I do still have some questions:
- Unusable rows from Pi, Tx, Exp or EVM circuits still look magical as there's no explanation of how these numbers were obtained or any rationale about why this is the number. (Not sure if it's desired to have, but would definitely make everything more clear).

File: zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs
Comment (han0110) [reply zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:25]: These number comes from different `num_rows_per_round` setting, still couldn't figure out a formula to define it, they are collected after running the `configure`.

**Comment by aguzmant103** [comment]
@CPerezz friendly ping that this PR is waiting for review :)

File: zkevm-circuits/src/exp_circuit.rs
Comment (ed255) [inline zkevm-circuits/src/exp_circuit.rs:507]:
```suggestion
        // Also column col2 and col3 of are queried at 8 distinct rotations at
```

**Review by ed255 — APPROVED** [review]
LGTM, modulo the typos and suggestion from David to use the constants in the EVM Circuit `unusable_rows`.
It's so nice to see all the "magic" numbers removed :D

File: zkevm-circuits/src/util.rs
Comment (han0110) [reply zkevm-circuits/src/util.rs:226]: It's actually from the comment of https://github.com/privacy-scaling-explorations/halo2/blob/b739b543d6f3d7d3a4ebb11660dd168136c4ae5f/halo2_proofs/src/plonk/circuit.rs#L2042-L2069 :)

File: zkevm-circuits/src/keccak_circuit/test.rs
Comment (han0110) [reply zkevm-circuits/src/keccak_circuit/test.rs]: Sounds good, added in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1282/commits/86b77ee9eb9f8d06a5459d8d330b62f1c33e062c

File: zkevm-circuits/src/util.rs
Comment (han0110) [reply zkevm-circuits/src/util.rs]: It's internal to halo2, it's already in upstream actually, so it's because we need to reserve an extra row for grand-product boundray check, so that row is not copy-able or lookup-able, I have added more comment about this extra one in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1282/commits/d54f0564ef955d5bf9c26407be80148fd53fa329

File: zkevm-circuits/src/evm_circuit.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit.rs]: Makes sense, updated in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1282/commits/d54f0564ef955d5bf9c26407be80148fd53fa329

**Pull request merged**
